funcx_endpoint/funcx_endpoint/endpoint/rabbit_mq/task_queue_subscriber.py

```diff
@@ -346,19 +349,30 @@ def close_channel(self):
         logger.info("Closing the channel")
         self._channel.close()
 
+    def _shutdown(self):
+        logger.debug("set status to 'closing'")
+        self.status = SubscriberProcessStatus.closing
+        logger.debug("closing connection")
+        self._connection.close()
+        logger.debug("stopping ioloop")
+        self._connection.ioloop.stop()
+        logger.debug("closing connection to mp queue")
+        self.external_queue.close()
```

Comment:
I thought that `ioloop.stop()` would guarantee that there's no concurrent call to `on_message`. I'll double-check the pika ioloop code to see about this. Otherwise, there needs to be a wait here, prior to closing the queue, to ensure the last `on_message` has processed.
An uncleanly closed queue was one of the things that I saw crash one of my local test runs.

Reply:
`ioloop.stop` documentation:
> Request exit from the ioloop. The loop is NOT guaranteed to stop before this method returns.

I really thought this would be safe at this point. There's also the fact that the connection is closed, so all of its channels are closed.

I'll see if the channel being closed guarantees that there's no `on_message` happening.

In the very worst case, I can add some kind of tracking for the message handling callbacks to ensure that we wait for the last message handler to run.

Reply:
Actually, this should be fine, if the put fails, we do not ack. As long as the function is not interrupted we are good.

Reply:
I read up on this, and `Channel.close()` can also enqueue things, rather than doing a synchronous close. I went ahead and added another event alongside `cleanup_complete` for `channel_closed` and have the `on_channel_closed` setting it.

I'll push the change. It also adds a warning log if we wait on these events and timeout while waiting, which I think is a pattern we should keep. You'll see it all in a second or two.